### PR TITLE
ddl: expend interface of lightning for add index acceleration

### PR DESCRIPTION
**Pull request by Benjamin2037**
### What problem does this PR solve?
Issue Number: ref #35983

Problem Summary:
This pr expand the currently interface for add index lightning solution.

### Check List

Tests 

- [X] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/engine.go [hunk 1]
```diff
@@ -235,6 +235,18 @@ func (e *Engine) unlock() {
 	e.mutex.Unlock()
 }
 
+func (e *Engine) TotalMemorySize() int64 {
+	var memSize int64 = 0
+	e.localWriters.Range(func(k, v interface{}) bool {
+		w := k.(*Writer)
+		if w.kvBuffer != nil {
+			memSize += w.kvBuffer.TotalSize()
```
Comment (gozssky) [inline br/pkg/lightning/backend/local/engine.go:243]: `kvBuffer` may not be a thread-safe object.

Comment (Benjamin2037) [reply br/pkg/lightning/backend/local/engine.go:244]: Added aquire/release writer lock to make it thread safe.

Comment (Benjamin2037) [reply br/pkg/lightning/backend/local/engine.go:244]: Added writer lock/unlock logic.

File: br/pkg/lightning/backend/backend.go [hunk 2]
```diff
@@ -215,6 +215,9 @@ type AbstractBackend interface {
 	// ResolveDuplicateRows resolves duplicated rows by deleting/inserting data
 	// according to the required algorithm.
 	ResolveDuplicateRows(ctx context.Context, tbl table.Table, tableName string, algorithm config.DuplicateResolutionAlgorithm) error
+
+	// Total Memory usage. This is only used for local backend
```
Comment (tangenta) [inline br/pkg/lightning/backend/backend.go:219]:
```suggestion
	// TotalMemoryConsume is the total memory usage. This is only used for local backend.
```

File: br/pkg/lightning/backend/backend.go [hunk 3]
```diff
@@ -411,11 +418,20 @@ func (engine *OpenedEngine) LocalWriter(ctx context.Context, cfg *LocalWriterCon
...
+func (engine *OpenedEngine) TotalMemoryConsume() int64 {
+	return engine.engine.backend.TotalMemoryConsume()
+}
+
 // WriteRows writes a collection of encoded rows into the engine.
 func (w *LocalEngineWriter) WriteRows(ctx context.Context, columnNames []string, rows kv.Rows) error {
 	return w.writer.AppendRows(ctx, w.tableName, columnNames, rows)
 }
 
+// WriteRows writes a collection of encoded rows into the engine.
```
Comment (tangenta) [inline br/pkg/lightning/backend/backend.go:430]:
```suggestion
// WriteRow writes an encoded row into the engine.
```

BTW, I don't understand why it is needed.

File: br/pkg/lightning/backend/kv/session_test.go [hunk 4]
```diff
@@ -1,29 +0,0 @@
-// Copyright 2019 PingCAP, Inc.
```
[the remark this answers is not in the document]
Comment (Benjamin2037) [reply br/pkg/lightning/backend/kv/session_test.go:1]: This is misstake to include another pr‘s code, I will refine the code.

File: br/pkg/lightning/backend/backend.go (hunk 2 above)
Comment (Benjamin2037) [reply br/pkg/lightning/backend/backend.go]: Done

File: br/pkg/lightning/backend/backend.go [hunk 5]
```diff
@@ -501,3 +512,7 @@ type EngineWriter interface {
 	IsSynced() bool
 	Close(ctx context.Context) (ChunkFlushStatus, error)
 }
+
+func (oe *OpenedEngine) GetEngineUuid() uuid.UUID {
+	return oe.uuid
+}
```
Comment (tangenta) [inline br/pkg/lightning/backend/backend.go:518]:
```suggestion
func (engine *OpenedEngine) GetEngineUuid() uuid.UUID {
	return engine.uuid
}
```

It is better to keep the name of receiver consistent.

**Review by tangenta — APPROVED** [review]

File: br/pkg/lightning/backend/local/local.go [hunk 6]
```diff
@@ -495,6 +495,18 @@ func NewLocalBackend(
...
+func (local *local) TotalMemoryConsume() int64 {
+	var memConsume int64 = 0
+	local.engines.Range(func(k, v interface{}) bool {
+		e := v.(*Engine)
+		if e != nil {
+			memConsume += e.TotalMemorySize()
+		}
+		return true
+	})
+	return memConsume
```
Comment (gozssky) [inline br/pkg/lightning/backend/local/local.go:507]: We should also consider the `memConsume` of `local.bufferPool`. When `kvBuffer` is destroyed, its memory is put back to the pool.

Comment (tangenta) [reply br/pkg/lightning/backend/local/local.go]: @gozssky Updated.

**Review by gozssky — APPROVED** [review]

**Pull request merged**